Simplenote/NoteEditorViewController+Swift.swift

```diff
@@ -831,6 +847,67 @@ extension NoteEditorViewController {
 }
 
 
+// MARK: - Shortcuts
+//
+extension NoteEditorViewController {
+    @objc
+    func toggleTagsAndEditor() {
+        if noteEditor.isFirstResponder {
+            view.window?.makeFirstResponder(tagsField)
+            tagsField.currentEditor()?.moveToEndOfDocument(nil)
+            tagsField.ensureCaretIsOnscreen()
+        } else {
+            view.window?.makeFirstResponder(noteEditor)
+        }
+    }
+}
+
+
+// MARK: - Editor Metadata
+//
+extension NoteEditorViewController {
+    @objc
+    func saveScrollPosition() {
+        // Issue #393: `self.note` might be populated, but it's simperiumKey inaccessible
+        guard let simperiumKey = note?.simperiumKey else {
+            return
+        }
+
+        let scrollPosition = scrollView.contentView.bounds.origin.y
+        metadataCache.store(scrollPosition: scrollPosition, for: simperiumKey)
+    }
+
+    @objc
+    func restoreScrollPosition() {
+        guard let simperiumKey = note?.simperiumKey,
+              let scrollPosition = metadataCache.metadata(for: simperiumKey)?.scrollPosition else {
+            scrollView.scrollToTop(animated: false)
+            return
+        }
+        // ensure layout to make sure that content size is updated
+        noteEditor.ensureLayout()
+        scrollView.documentView?.scroll(NSPoint(x: 0, y: scrollPosition))
+    }
+
+    @objc
+    func saveCursorLocation() {
+        guard let simperiumKey = note?.simperiumKey else {
+            return
+        }
+        metadataCache.store(cursorLocation: noteEditor.selectedRange().location, for: simperiumKey)
+    }
+
+    @objc
+    func restoreCursorLocation() {
+        guard let simperiumKey = note?.simperiumKey,
+              let cursorLocation = metadataCache.metadata(for: simperiumKey)?.cursorLocation else {
+            return
+        }
+        noteEditor.setSelectedRange(NSRange(location: cursorLocation, length: 0))
```

Comment:
Just noting that (I've just verified!!) that if the Cursor Location isn't really valid, NSTextView doesn't crash.

(I was thinking: what if there are unaccounted scenarios that could render invalid the old Location)